### PR TITLE
RFCs: Alter Primary Key

### PR DESCRIPTION
**Pull request by dt**
Document the proposed process for changing primary keys.

Release note: none.

### CODE REVIEW

**Comment by RaduBerinde** [comment]
This is pretty clever. But I think it will be hard to make it work for non-unique indexes. Say my PK is `x`. A non-unique index appends `x` to each key (in fact, even a unique index does that for NULL values). If I change my PK to `y`, will we continue to use `x` as the extra key in the index? In which case, what prevents me from removing the old PK index and then adding duplicates on column `x`?

**Comment by dt** [comment]
@RaduBerinde Yeah, those using the old STORING format I think work as-is, since both `x` and `y` (assuming y is stored, and x is the old PK`) would be encoded in the key. For the new `STORING` format, `y` would be in the value.

Initially I was thinking I'd just make you replace that index with one that stores both `x` and `y` in the keys before the swap -- index desc will need to start explicitly listing stored-in-key and stored-in-value columns anyway (rather than inferring the extra ones based on the PK, since the index on `x` still stores everything after it stops being the PK. Indeed, I think this will actually generalize some index selection stuff elsewhere, since we can remove special-cases for PK and just go by whats in the desc).

As long as the old PK index is kept around as a unique index though, I guess you could get away with `y` staying in the value half, since the uniqueness of `x` means it is still sufficient to make unique index keys. This is pretty special-case-y and would complicate the flow-chart for index encoding substantially, so i'm not inclined to do it, but could save an operator some index-rewrites so if user feedback suggests it would be a big deal, maybe we decide it is worth it?

**Comment by knz** [comment]
We talked about this in-person. I am thrilled by the elegance of the scheme. No further comment at this point than the points already raised by Radu.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[docs/RFCS/20180413_alter_primary_key.md, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25208#-LBWGkGMDKAeB6KqbpwC:-LBWGkGO1hLyTwaMyC4U:b12yylh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0a64b241330ec426d213a0c776a49a9f1fc14a7/docs/RFCS/20180413_alter_primary_key.md#L120)):*
> ```Markdown
> 
> Thus instead of `ADD` we will use `ALTER` in to edit the existing key, e.g.:
> `ALTER PRIMARY KEY (col1, col2) USING INDEX idx_foo`
> ```

nit: `ALTER TABLE ... PRIMARY KEY ...`

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25208#-:-LBWGtrt4iKUnNlFOZp-:b-4xh4nh)*

**Comment by danhhz** [comment]
I like this direction! Much better than what I was thinking we'd have to do.

I like to see a discussion added of how column families work with this. Specifically, say for example that you have a table with 2 column families and want to alter the primary key while keeping the same column families. Unless something has changed since I looked, secondary indexes never have column families, so it's currently not possible to construct the index you'd swap to.

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[docs/RFCS/20180413_alter_primary_key.md, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25208#-LBWpQzY40OVamDHxIaS:-LBWpQzY40OVamDHxIaT:bwzs839) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0a64b241330ec426d213a0c776a49a9f1fc14a7/docs/RFCS/20180413_alter_primary_key.md#L19)):*
> ```Markdown
> # Summary
> 
> Users should be able to change the primary key of an existing table.
> ```

Does this include being able to change the interleaving? (I guess it should work but would need new syntax)

---

*[docs/RFCS/20180413_alter_primary_key.md, line 94 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25208#-LBWqGKy36gSxHwm6Siv:-LBWqGKy36gSxHwm6Siw:b-eary5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0a64b241330ec426d213a0c776a49a9f1fc14a7/docs/RFCS/20180413_alter_primary_key.md#L94)):*
> ```Markdown
> 
> Later improvements may include:
>   1) add the ability to implicitly add the new index that will be marked primary
> ```

And the ability to automatically drop the old PK index. 

---

*[docs/RFCS/20180413_alter_primary_key.md, line 117 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25208#-LBWqcQsDdEAewXFwBXN:-LBWqcQsDdEAewXFwBXO:b-5n3onu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0a64b241330ec426d213a0c776a49a9f1fc14a7/docs/RFCS/20180413_alter_primary_key.md#L117)):*
> ```Markdown
> workaround for tables that do not specify a primary key is to add and index an 
> extra column, making such a `DROP` a relatively expensive operation that we are
> unlikely to implement.
> ```

Adding a primary key to a table that didn't have one would leave the table with a vestigial hidden `rowid` column (stored in all secondary indexes as of the completion of the schema change). 

Ideally the `rowid` column should eventually disappear without a trace, although rewriting all secondary indexes and dropping that column would be expensive.

---

*[docs/RFCS/20180413_alter_primary_key.md, line 152 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25208#-LBWuIuN1DSpCfpDc8U8:-LBWuIuN1DSpCfpDc8U9:b6qm0qi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0a64b241330ec426d213a0c776a49a9f1fc14a7/docs/RFCS/20180413_alter_primary_key.md#L152)):*
> ```Markdown
> ### Special Handling of Primary Indexes
> 
> In various places, the descriptor handling of the primary index is special-cased
> ```

We also need to do some compatibility research around index naming. The primary key is named "primary" by default, although it is possible to rename it. Should we rename indexes when changing primary keys, or leave things in the somewhat confusing state of a secondary index named "primary" and a primary key named something else?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25208)*

**Comment by dt** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/RFCS/20180413_alter_primary_key.md, line 152 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25208#-LBWuIuN1DSpCfpDc8U8:-LBWy6DB-QigrVq1-W6F:b-6v2l2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0a64b241330ec426d213a0c776a49a9f1fc14a7/docs/RFCS/20180413_alter_primary_key.md#L152)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We also need to do some compatibility research around index naming. The primary key is named "primary" by default, although it is possible to rename it. Should we rename indexes when changing primary keys, or leave things in the somewhat confusing state of a secondary index named "primary" and a primary key named something else?
</blockquote></details>

I think we should _not_ automatically rename since queries can reference indexes by name, and I think we should assume they're still referencing the same index.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25208)*

File: docs/RFCS/20180413_alter_primary_key.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+supported in-place alteration.
+
+### Compatible Primary Indexes
+
+The definition of primary key -- `UNIQUE` and `NOT NULL` adds some constraints
+on which indexes are eligible to become the primary key:
+  1) The index must be `UNIQUE`, indexing _exactly_ the columns intended to be
+     the new primary key (as that set is what is constrained to be unique). must
+     be unique and its columns must be `NOT NULL`.
+  2) All indexed columns must be `NOT NULL`
```
Comment (mjibson) [inline docs/RFCS/20180413_alter_primary_key.md]: this point is repeated at the end of the proceeding bullet

**Comment by dt** [comment]
This feature got de-prioritized by the bulk io and schema change team, in favor of putting resources elsewhere on the roadmap for the near term.

Our current thinking is to merge this doc in draft status, so it can serve as a starting point for when we pick it back up. My plan to try to try to capture most of the feedback and comments made here as  TODOs in the document itself, stating what needs to be clarified/expanded/etc before the draft can be accepted, but wanted to check if anyone has other opinions/preferences before going that route.

**Comment by RaduBerinde** [comment]
Sounds good to me.

**Comment by dt** [comment]
Added some TODOs and an explicit note about draft/paused status. RFAL.

**Review by madelynnblue — APPROVED** [review]

**Comment by dt** [comment]
whoops, forgot to actually merge this after adding the TODOs and draft status marker.

bors r+

**Pull request merged**